cmd/server/main.go

```diff
@@ -187,96 +190,82 @@ func bundle(w http.ResponseWriter, r *http.Request) {
-			err = errors.Wrapf(err, "error parsing override image %q: %v", image, err)
-			handleHttpError(w, r, err, http.StatusInternalServerError)
-			return
-		}
-
-		tempDir, err := ioutil.TempDir("/images", "temp-image-pull")
+	var tempDir string
+	if len(bundle.Images) > 0 {
+		tempDir, err = ioutil.TempDir("/images", "temp-image-pull")
 		if err != nil {
 			err = errors.Wrap(err, "error creating temp directory")
-			handleHttpError(w, r, err, http.StatusInternalServerError)
+			handleError(r.Context(), err)
 			return
 		}
 		defer os.RemoveAll(tempDir)
+	}
 
-		destPath := path.Join(tempDir, "temp-archive-image")
-		destStr := fmt.Sprintf("docker-archive:%s:%s", destPath, image)
-		localRef, err := alltransports.ParseImageName(destStr)
-		if err != nil {
-			err = errors.Wrapf(err, "failed to parse local image name: %s", destStr)
-			handleHttpError(w, r, err, http.StatusInternalServerError)
-			return
-		}
-
-		policy, err := signature.NewPolicyFromBytes(imagePolicy)
-		if err != nil {
-			err = errors.Wrap(err, "failed to read default image policy")
-			handleHttpError(w, r, err, http.StatusInternalServerError)
-			return
-		}
-		policyContext, err := signature.NewPolicyContext(policy)
-		if err != nil {
-			err = errors.Wrap(err, "failed to create image policy context")
-			handleHttpError(w, r, err, http.StatusInternalServerError)
-			return
-		}
-		destCtx := &types.SystemContext{
-			DockerDisableV1Ping: true,
-		}
-		srcCtx := &types.SystemContext{
-			DockerDisableV1Ping: true,
-			AuthFilePath:        "/dev/null",
-			DockerAuthConfig: &types.DockerAuthConfig{
-				Username:      "",
-				Password:      "",
-				IdentityToken: "",
-			},
-		}
-		_, err = copy.Image(r.Context(), policyContext, localRef, srcRef, &copy.Options{
-			RemoveSignatures:      true,
-			SignBy:                "",
-			ForceManifestMIMEType: "",
-			DestinationCtx:        destCtx,
-			SourceCtx:             srcCtx,
-		})
-		if err != nil {
-			err = errors.Wrapf(err, "failed to save docker image archive of %s", image)
-			handleHttpError(w, r, err, http.StatusInternalServerError)
-			return
-		}
-
-		f, err := os.Open(destPath)
-		if err != nil {
-			err = errors.Wrap(err, "failed to open override image archive")
-			handleHttpError(w, r, err, http.StatusInternalServerError)
-			return
-		}
-		fi, err := f.Stat()
-		if err != nil {
-			handleHttpError(w, r, err, http.StatusInternalServerError)
-			return
-		}
-
-		header := &tar.Header{
-			Name:    fmt.Sprintf("kurl/image-overrides/%d.tar", i),
-			Size:    fi.Size(),
-			Mode:    0644,
-			ModTime: time.Now(),
+	errCh := make(chan error, 1)
+	defer close(errCh)
+
+	imageCh := make(chan string, len(bundle.Images))
+
+	// HACK: Work around the fact that github.com/containers/image/v5/copy copy.Image does not support streaming.
+	// Download images concurrently while streaming the rest of the bundle to minimize idle time.
+	go func() {
+		defer close(imageCh)
+
+		for i, image := range bundle.Images {
```

Comment:
Based on the title of this PR, wouldn't you want this for loop outside the goroutine, to download concurrently?

Reply:
It will download them sequentially but concurrently while streaming the rest of the bundle